src/constraints/RebarBondSlipConstraint.C

```diff
@@ -0,0 +1,517 @@
+  MooseEnum bondslip_model("concrete_rebar_model elastic_perfect_plastic_model",
+                           "elastic_perfect_plastic_model");
+  params.addParam<MooseEnum>("bondslip_model", bondslip_model, "Node-Element bond slip model.");
+
+  params.addCoupledVar("output_axial_slip", "Rebar axial slip output variable.");
+  params.addCoupledVar("output_axial_force", "Rebar axial force output variable");
+  params.addCoupledVar("output_axial_plastic_slip", "Rebar axial plastic component of bond slip");
+  return params;
+}
+
+template <bool is_ad>
+RebarBondSlipConstraintTempl<is_ad>::RebarBondSlipConstraintTempl(
+    const InputParameters & parameters)
+  : EqualValueEmbeddedConstraintTempl<is_ad>(parameters),
+    _bondslip_model(
+        this->template getParam<MooseEnum>("bondslip_model").template getEnum<BondSlipModel>()),
+    _component(libMesh::invalid_uint),
+    _mesh_dimension(_mesh.dimension()),
+    _disp_vars_nums(_mesh_dimension, libMesh::invalid_uint),
+    _disp_vars(_mesh_dimension, nullptr),
+    _max_bondstress(this->template getParam<Real>("max_bondstress")),
+    _bar_radius(this->template getParam<Real>("rebar_radius")),
+    _transitional_slip(this->template getParam<Real>("transitional_slip_value"))
+{
+  if (_mesh_dimension != this->coupledComponents("displacements"))
+    mooseError("In RebarBondSlipConstraint, number of displacements must equal the mesh dimension");
+
+  for (unsigned int i = 0; i < _mesh_dimension; ++i)
+  {
+    _disp_vars_nums[i] = this->coupled("displacements", i);
+    _disp_vars[i] = this->getVar("displacements", i);
+  }
+
+  for (std::size_t i = 0; i < _disp_vars_nums.size(); ++i)
+    if (_var.number() == _disp_vars_nums[i])
+      _component = i;
+  if (_component == libMesh::invalid_uint)
+    mooseError("Problem with displacements in " + this->_name);
+
+  if (this->isParamValid("output_axial_slip"))
+    _output_axial_slip = &(this->writableVariable("output_axial_slip"));
+  if (this->isParamValid("output_axial_force"))
+    _output_axial_force = &(this->writableVariable("output_axial_force"));
+  if (this->isParamValid("output_axial_plastic_slip"))
+    _output_axial_plastic_slip = &(this->writableVariable("output_axial_plastic_slip"));
+}
+
+template <bool is_ad>
+void
+RebarBondSlipConstraintTempl<is_ad>::initialSetup()
+{
+
+  for (auto it = _secondary_to_primary_map.begin(); it != _secondary_to_primary_map.end(); ++it)
+    if (_bondslip.find(it->first) == _bondslip.end())
+      _bondslip.insert(std::pair<dof_id_type, bondSlipData>(it->first, bondSlipData()));
+}
+
+template <bool is_ad>
+void
+RebarBondSlipConstraintTempl<is_ad>::timestepSetup()
+{
+  for (auto iter = _secondary_to_primary_map.begin(); iter != _secondary_to_primary_map.end();
+       ++iter)
+  {
+    dof_id_type node_id = iter->first;
+    [[maybe_unused]] auto it = _bondslip.find(node_id);
+    mooseAssert(it != _bondslip.end(), "Node not found in bond-slip map");
```

Comment:
```suggestion
    mooseAssert(bondslip.find(node_id) != _bondslip.end(), "Node not found in bond-slip map");
```

Reply:
I'm suggesting that change because the expensive part is the `find()`, so that should be inside the assert if possible.

Also, I'm thinking of when this might come up, and it occurred to me that we would need to re-do searches when the mesh changes. Maybe that happens already in the base classs? We should check.